src/XML/SignedElementInterface.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+declare(strict_types=1);
+
+namespace SimpleSAML\XMLSecurity\XML;
+
+use SimpleSAML\XMLSecurity\XMLSecurityKey;
+
+/**
+ * An interface describing signed elements.
+ *
+ * @package simplesamlphp/xml-security
+ */
+interface SignedElementInterface
+{
+    /**
+     * Retrieve certificates that sign this element.
+     *
+     * @return array Array with certificates.
+     * @throws \Exception if an error occurs while trying to extract the public key from a certificate.
+     */
+    public function getValidatingCertificates(): array;
+
+
+    /**
+     * Validate this element against a public key.
+     *
+     * If no signature is present, false is returned. If a signature is present,
+     * but cannot be verified, an exception will be thrown.
+     *
+     * @param \SimpleSAML\XMLSecurity\XMLSecurityKey $key The key we should check against.
+     * @return bool True if successful, false if we don't have a signature that can be verified.
+     */
+    public function validate(XMLSecurityKey $key): bool;
```

Comment:
I think, in order to encourage good practices, and also to avoid any possible implementation errors (e.g. we have a signed element, and we just call `getElement()` to get the unsigned element without validating the signature first), the `validate()` method should return the actual element whose signature has been verified.

That would mean that when the signature validation fails, we should throw an exception (which is a good way to ensure the failure is noticeable, I believe), instead of returning false. It would also prevent us from having any processing security issue. We'll need support from xmlseclibs or our fork though to get this done, since verify should call `fromXML()` on the target class (the signable element), passing the actual, canonicalised an transformed DOM (not the original), and as far as I recall, that's not possible with xmlseclibs.